### PR TITLE
Add spawnProcessDetached

### PR DESCRIPTION
**Pull request by FreeSlave**
spawnProcessDetached is function similar to spawnProcess, but user does not need to wait on spawned process to avoid zombies (resource leak). Instead of returning Pid instance spawnProcessDetached returns integer pid via pointer. Rationale: user should not be able to wait on detached process.

Currently only Posix version is implemented. It uses double fork technique and setsid to detach the new process from its caller. Since this implementation reuses a lot of stuff from spawnProcess I moved some implementation details to separate private functions to share them between spawnProcess and spawnProcessDetached.

THIS IS WIP
I'll add more unittests and Windows version later. For now say if you like the idea.

### CODE REVIEW

**Comment by CyberShadow** [comment]
I believe that on Windows, you simply have to close the process handle as soon as the process is created to effectively "detach" the process. (This is not the same as a detached console window.)

Can this be implemented as a `Config` flag instead? I'm not sure about adding an entire overload. Then it can be used for `spawnShell` as well. `Pid` will likely need a few changes to make waiting on a detached processes an explicit error.

**Comment by FreeSlave** [comment]
> I believe that on Windows, you simply have to close the process handle as soon as the process is created to effectively "detach" the process.

I believe so. This is another reason why I did not want to use Pid class. Process handle is part of Pid class on Windows. Closing it will make osHandle be able to be null. So users will need to check for osHandle not being null if they want to use it. That differs from the current situation when osHandle is always not null. And it's also inconsistent with Posix, where osHandle is same as pid and therefore always valid. Do you think it is ok?

**Comment by CyberShadow** [comment]
> So users will need to check for osHandle not being null if they want to use it.

I think it's more of not doing a thing that doesn't make sense to do.

> And it's also inconsistent with Posix, where osHandle is same as pid and therefore always valid.

If you spawn a detached process, technically there is no guarantee that the PID remains valid for any amount of time - the process may exit on its own and the OS may assign the PID to another process. So, if we are to reuse `Pid`, probably the best course is a private `bool owned` field (which is false for detached processes), and `kill` and `wait` require that `owned` is true.

CC @schveiguy @kyllingstad

**Comment by FreeSlave** [comment]
> kill and wait require that owned is true

Why would kill require owned is true? You can kill any process via task manager or kill program which are clearly does not own this process.

**Comment by CyberShadow** [comment]
Because it might not be the process you actually intend to kill - that process may have exited a long time ago.

**Comment by FreeSlave** [comment]
Got it.
I'm still not sure that Pid class can be re-used here. There was that PR https://github.com/dlang/phobos/pull/5086 that introduced possibility to have non-owned Pid instances. And it was reverted later. Probably because it complicates the usage of other functions (like already said `wait` and `kill`).

**Comment by CyberShadow** [comment]
#5086 had more problems than that, though (killing processes being an intrinsically platform-dependent operation; needing to call `OpenProcess` on Windows). I think overloading `Pid` here makes more sense than adding an overload set.

**Comment by kyllingstad** [comment]
> I think overloading `Pid` here makes more sense than adding an overload set.

I very much agree with this.

**Comment by FreeSlave** [comment]
Added Pid.owned propert and Config.detached constant.
Also added separate unittest to check non-ability of waiting/killing not owned process. It runs OK, but I have this message at terminal
```
/bin/sh: 0: Can't open /tmp/std.process temporary file 76b5fac2-57f8-4f5b-88dc-9fd27c32ae73
```

File: std/process.d [hunk 1]
```diff
@@ -338,7 +339,131 @@ version(Posix) private enum InternalError : ubyte
...
     chdir,
-    getrlimit
+    getrlimit,
+    doubleFork
+}
+
+version(Posix)
+{
+private:
+    const(char)[] getExecutablePath(const(char)[] name)
```
Comment (CyberShadow) [inline std/process.d]: Since there is no longer an overload set, these refactorings aren't needed any more, right? If so, can they be undone?

File: std/process.d [hunk 2]
```diff
@@ -1316,6 +1428,12 @@ enum Config
     process.
     */
     inheritFDs = 32,
+
+    /**
+    Spawn process in detached state. This removes the need in calling
+    $(LREF wait) to clean up the process resources.
```
Comment (CyberShadow) [inline std/process.d]: Should have a note here that calling `wait` or `kill` with the resulting `Pid` is invalid.

File: std/process.d [hunk 3]
```diff
@@ -1358,6 +1476,15 @@ final class Pid
         return _processID;
     }
 
+    /**
+    This is $(D true) if process was spawned without $(LREF Config.detached), $(D false)
+    otherwise. You can't $(LREF wait) for or $(LREF kill) process you don't own.
+    */
+    @property bool owned() const @safe pure nothrow
```
Comment (CyberShadow) [inline std/process.d]: No need to make this public. I can't think of many cases that programs would mix owned and non-owned Pids and require distinguishing them. Making at an implementation detail also allows changing the underlying implementation in the future, if necessary.

File: std/process.d [hunk 4]
```diff
@@ -1374,6 +1501,8 @@ private:
     version (Posix)
     int performWait(bool block) @trusted
     {
+        import std.exception : enforceEx;
+        enforceEx!ProcessException(_owned, "Attempt to wait for the not owned process");
```
Comment (CyberShadow) [inline std/process.d]: Consider "Can't wait on a detached process"

File: std/process.d [hunk 5]
```diff
@@ -1422,6 +1551,8 @@ private:
     {
         int performWait(bool block) @trusted
         {
+            import std.exception : enforceEx;
+            enforceEx!ProcessException(_owned, "Attempt to wait for the not owned process");
```
Comment (CyberShadow) [inline std/process.d]: Ditto

File: std/process.d [hunk 6]
```diff
@@ -1503,7 +1647,7 @@ Signal codes are defined in the $(D core.sys.posix.signal) module
 (which corresponds to the $(D signal.h) POSIX header).
 
 Throws:
-$(LREF ProcessException) on failure.
+$(LREF ProcessException) on failure or on attempt to wait for not owned process.
```
Comment (CyberShadow) [inline std/process.d]: Consider "detached" or "non-child" instead of "not owned", the "owned" term should be an implementation detail

File: std/process.d [hunk 7]
```diff
@@ -1558,7 +1702,7 @@ Returns:
 An $(D std.typecons.Tuple!(bool, "terminated", int, "status")).
 
 Throws:
-$(LREF ProcessException) on failure.
+$(LREF ProcessException) on failure or on attempt to wait for not owned process.
```
Comment (CyberShadow) [inline std/process.d]: ditto

File: std/process.d [hunk 8]
```diff
@@ -1636,6 +1780,7 @@ assert(wait(pid) == -SIGKILL); // Negative return value on POSIX!
 
 Throws:
 $(LREF ProcessException) on error (e.g. if codeOrSignal is invalid).
+    or on attempt to kill not owned process.
```
Comment (CyberShadow) [inline std/process.d]: ditto

File: std/process.d [hunk 9]
```diff
@@ -1652,6 +1797,8 @@ void kill(Pid pid)
 /// ditto
 void kill(Pid pid, int codeOrSignal)
 {
+    import std.exception : enforceEx;
+    enforceEx!ProcessException(pid.owned(), "Attempt to kill not owned process");
```
Comment (CyberShadow) [inline std/process.d]: ditto

**Comment by CyberShadow** [comment]
> It runs OK, but I get this message at terminal

Is this an expected and unavoidable effect of the unit test, or do you need time/assistance in tracking it down?

**Comment by FreeSlave** [comment]
> Is this an expected and unavoidable effect of the unit test, or do you need time/assistance in tracking it down?

It should not happen. It's simple TestScript usage. But I can't understand why it happens.

**Comment by FreeSlave** [comment]
Ok, I think I got it. TestScript destructor runs (effectively removing the file) before file execution. But kernel somehow manages to find out that file should have ran with /bin/sh. Weird.

File: std/process.d [hunk 10]
```diff
@@ -416,15 +418,21 @@ private Pid spawnProcessImpl(in char[][] args,
...
 
+    int[2] pidPipe;
+    if (config & Config.detached)
+    {
+        if (core.sys.posix.unistd.pipe(pidPipe) == 0)
+            setCLOEXEC(pidPipe[1], true);
+        else
+            throw ProcessException.newFromErrno("Could not create pipe to get process pid");
+    }
+    scope(exit) if (config & Config.detached) close(pidPipe[0]);
```
Comment (CyberShadow) [inline std/process.d]: Any reason to not reuse the same pipe used for communicating the result (`forkPipe`)?

File: std/process.d [hunk 11]
```diff
@@ -450,94 +465,133 @@ private Pid spawnProcessImpl(in char[][] args,
...
+                if (fchdir(workDirFD) < 0)
+                {
+                    // Fail. It is dangerous to run a program
+                    // in an unexpected working directory.
+                    abortOnError(forkPipeOut, InternalError.chdir, .errno);
+                }
+                close(workDirFD);
             }
-            close(workDirFD);
         }
```
Comment (CyberShadow) [inline std/process.d]: Why change the working directory in the second fork? Wouldn't it be simpler to change it in the first fork regardless of `Config.detached`?

File: std/process.d [hunk 12]
```diff
@@ -450,94 +465,133 @@ private Pid spawnProcessImpl(in char[][] args,
...
+
+        if (config & Config.detached)
+        {
+            auto secondFork = core.sys.posix.unistd.fork();
+            if (secondFork == 0)
             {
-                pfds[i].fd = i + 3;
-                pfds[i].events = 0;
-                pfds[i].revents = 0;
+                setsid();
```
Comment (CyberShadow) [inline std/process.d]: Hmm, isn't `setsid` supposed to be called from the first fork (so that the session leader is killed)?

File: std/process.d [hunk 13]
```diff
@@ -1359,6 +1459,11 @@ final class Pid
     }
 
 private:
+    @property bool owned() const @safe pure nothrow
```
Comment (CyberShadow) [inline std/process.d]: No point in a private property, just name the field `owned`.

File: std/process.d [hunk 14]
```diff
@@ -1709,6 +1834,25 @@ void kill(Pid pid, int codeOrSignal)
...
+    import std.exception : assertThrown;
+    TestScript prog = "exit 0";
+    auto pid = spawnProcess([prog.path], null, Config.detached);
+    /*
+    This sleep is needed because we can't wait() for detached process to end
+    and therefore TestScript destructor may run at the same time as /bin/sh tries to start the script.
+    This leads to the annoying message like "/bin/sh: 0: Can't open /tmp/std.process temporary file" to appear when running tests.
+    It does not happen in unittests with non-detached processes because we always wait() for them to finish.
+    */
+    Thread.sleep(dur!"msecs"(5));
```
Comment (CyberShadow) [inline std/process.d]: I think this is fine, but I'd give it a more generous timeout of 100 msecs or so just to exclude race conditions.

File: std/process.d (hunk 12 above)
Comment (FreeSlave) [inline std/process.d]: You're right. Will change.

File: std/process.d (hunk 14 above)
Comment (FreeSlave) [inline std/process.d]: I used 5 msecs, because there's the same sleep in the other unittest under version(Android), having a comment about Android killing sleeping applications.

Comment (CyberShadow) [inline std/process.d]: That's fine, we don't have CI for Android; I'm more concerned about Windows, where process creation is very slow. If the Android maintainer finds that this test is causing them trouble then they can fix it as needed. (Also, I was pretty sure Android kills idle apps, not processes...)

File: std/process.d (hunk 10 above)
Comment (FreeSlave) [inline std/process.d]: After the second fork started the first and the second forks run in parallel. I can't ensure that pid will be written to the pipe before possible errors.

Comment (CyberShadow) [inline std/process.d]: I see. Please add this explanation as a comment somewhere.

Comment (CyberShadow) [inline std/process.d]: I guess calling `getpid` from the second child is also a possibility, but I guess it's not worth the change.

Comment (FreeSlave) [inline std/process.d]: There's also a possibility that the second fork fails. In this case we would need to pass fictitious pid from the first fork to adjust.

File: std/process.d (hunk 12 above)
Comment (FreeSlave) [inline std/process.d]: I found that setsid makes some apps run wrong when starting from terminal without setting stdin to /dev/null.
E.g. run `setsid openarena` from terminal. It shows a window, but does render anything.
Should we do anything about that?

Comment (CyberShadow) [inline std/process.d]: Unfortunately I am not sufficiently familiar with the subject to confidently answer that.

From looking online, I noticed that `setsid` is something that is often done when double-forking, but not prescriptively so. Perhaps it only makes sense for non-interactive daemon processes?

Comment (FreeSlave) [inline std/process.d]: If we don't setsid, the spawned process will belong to the same session so pkill -s '$sid' will send signal to this process too.

Comment (CyberShadow) [inline std/process.d]: When is that a problem?

Comment (FreeSlave) [inline std/process.d]: Oh, well, not all apps belong to the same session actually. Only those that was ran from desktop. If it runs from terminal, session is different.
The point is that user may want to have it in another session or in the same. I think it's better to leave to user to decide.

Comment (CyberShadow) [inline std/process.d]: OK, so correct me if I'm wrong, but as I understand the most commonly-observed difference will be that ^C will not kill the spawned process if it's in the same session.

In which case, it does sound like it's worthy of a `Flag` addition, with the Windows implementation being mapped to the `CREATE_NEW_PROCESS_GROUP` process creation flag.

Comment (CyberShadow) [inline std/process.d]: It may be worth leaving it for a separate PR.

Comment (FreeSlave) [inline std/process.d]: ^C is sent only to the foreground processes of session. And session has distinction between foreground and background processes only if it has connection to controlling terminal.
Even if doing setsid via flag it's still not clear if user want the new process to be a session leader (i.e. call setsid in the second fork) or not (call setsid in the first fork and let it die).

> It may be worth leaving it for a separate PR.
So shall I remove setsid for now?

Comment (CyberShadow) [inline std/process.d]: Thanks for the clarification. I think I'm out of my depth here. Perhaps someone more knowledgeable than me about POSIX process sessions can chime in, @kyllingstad @schveiguy @jmdavis ?

Comment (FreeSlave) [inline std/process.d]: I've done some experiments. Login via ssh (just plain login without tmux, screen or similar) and spawned some script process in detached state without calling to setsid. Logout and re-login again, the process still ran. So I guess setsid is not necessary after all.

**Comment by CyberShadow** [comment]
Could you please rebase on top of the latest master? That should fix CircleCI.

**Comment by FreeSlave** [comment]
Rebased. Is any other change needed?

**Comment by CyberShadow** [comment]
Ah, a changelog entry or linked bug as per @dlang-bot's comment. I think that's all if there is no further feedback.

**Comment by FreeSlave** [comment]
I've added changelog https://github.com/FreeSlave/phobos/commit/fa8a73f47b953049ee8751bffb31401d33d79344, but github does not see the changes. Probably because of the previous rebase.

**Comment by CyberShadow** [comment]
@wilzbach Any idea why changelog is not showing up in DAutoTest diff here?

**Comment by wilzbach** [comment]
> @wilzbach Any idea why changelog is not showing up in DAutoTest diff here?

Yes, it seems like Martin has added this file and thus it doesn't get generated:

https://github.com/dlang/dlang.org/blob/master/changelog/2.075.0_pre.dd

Should we bump the version at dlang.org? Probably a good idea to do this for all version files after the master -> stable branch-off?

**Comment by CyberShadow** [comment]
> Should we bump the version at dlang.org? Probably a good idea to do this for all version files after the master -> stable branch-off?

Sounds like a good idea.

File: changelog/std-process-config-detached.dd [hunk 15]
```diff
@@ -0,0 +1,5 @@
+`Config.detached` flag for `spawnProcess` has been added
+
+Config.detached allows to $(REF_ALTTEXT spawn process, spawnProcess, std, process) independently from the caller.
+No need to wait on the process spawned that way - no zombie will left after process exits.
+Attempts to $(REF wait, std, process) on or $(REF kill, std, process) detached processes will throw.
```
Comment (CyberShadow) [inline changelog/std-process-config-detached.dd]: Some copyediting:

`Config.detached` allows $(REF_ALTTEXT spawning processes, spawnProcess, std, process) which run independently from the current process. There is no need to `wait` on the processes created with this flag, and no $(HTTPS en.wikipedia.org/wiki/Zombie_process, zombie process) will be left after they exit.
Attempts to call $(REF wait, std, process) or $(REF kill, std, process) on detached processes will throw.

**Comment by JackStouffer** [comment]
Does this really need Andrei's approval? It doesn't really add a new symbol.

**Comment by wilzbach** [comment]
> Does this really need Andrei's approval? It doesn't really add a new symbol.

Well there's `Config.detached`, but yeah I'm a bit frustrated about the huge bottleneck that this is causing as well. Maybe
- we should let the bot send @andralex a mail every week about open PRs which require his approval
- move away from this single point of failure and redefine this policy to "at least two codeowners need to approve a new symbol" (or require the @andralex label only for modules he cares about, i.e. for which he is a code owner)

File: std/process.d [hunk 16]
```diff
@@ -339,7 +340,8 @@ version(Posix) private enum InternalError : ubyte
     noerror,
     exec,
     chdir,
-    getrlimit
+    getrlimit,
+    doubleFork
```
Comment (andralex) [inline std/process.d]: add a comma so the next addition blames properly

File: std/process.d [hunk 17]
```diff
@@ -415,17 +417,30 @@ private Pid spawnProcessImpl(in char[][] args,
...
-    {
         throw ProcessException.newFromErrno("Could not create pipe to check startup of child");
-    }
     scope(exit) close(forkPipe[0]);
 
+    // To create detached process, we use double fork technique
+    // but we don't have a direct access to the second fork pid from the caller side thus use a pipe.
+    // We also can't reuse forkPipe for that purpose
+    // because we can't predict the order in which pid and possible error will be written
+    // since the first and the second forks will run in parallel.
```
Comment (andralex) [inline std/process.d]: looks like a multiline comment

File: std/process.d [hunk 18]
```diff
@@ -415,17 +417,30 @@ private Pid spawnProcessImpl(in char[][] args,
...
+    int[2] pidPipe;
+    if (config & Config.detached)
+    {
+        if (core.sys.posix.unistd.pipe(pidPipe) == 0)
+            setCLOEXEC(pidPipe[1], true);
+        else
+            throw ProcessException.newFromErrno("Could not create pipe to get process pid");
```
Comment (andralex) [inline std/process.d]: Simplify flow:

```d
if (core.sys.posix.unistd.pipe(pidPipe) != 0)
    throw ...;
setCLOEXEC(pidPipe[1], true);
```

File: std/process.d (hunk 10 above)
Comment (andralex) [inline std/process.d]: make these considerations a nice comment

File: std/process.d [hunk 19]
```diff
@@ -465,80 +490,110 @@ private Pid spawnProcessImpl(in char[][] args,
...
+        void execProcess()
         {
-            import core.stdc.stdlib : malloc;
-            import core.sys.posix.poll : pollfd, poll, POLLNVAL;
-            import core.sys.posix.sys.resource : rlimit, getrlimit, RLIMIT_NOFILE;
+            // Redirect streams and close the old file descriptors.
+            // In the case that stderr is redirected to stdout, we need
+            // to backup the file descriptor since stdout may be redirected
+            // as well.
+            if (stderrFD == STDOUT_FILENO)  stderrFD = dup(stderrFD);
```
Comment (andralex) [inline std/process.d]: double space?

File: std/process.d [hunk 20]
```diff
@@ -465,80 +490,110 @@ private Pid spawnProcessImpl(in char[][] args,
...
+                immutable maxDescriptors = cast(int) r.rlim_cur;
 
-            // Get the maximum number of file descriptors that could be open.
-            rlimit r;
-            if (getrlimit(RLIMIT_NOFILE, &r) != 0)
+                // The above, less stdin, stdout, and stderr
+                immutable maxToClose = maxDescriptors - 3;
+
+                // Call poll() to see which ones are actually open:
+                pollfd* pfds = cast(pollfd*) malloc(pollfd.sizeof * maxToClose);
```
Comment (andralex) [inline std/process.d]: auto

File: std/process.d [hunk 21]
```diff
@@ -548,18 +603,29 @@ private Pid spawnProcessImpl(in char[][] args,
     }
     else
     {
-        close(forkPipe[1]);
+        closePipeWriteEnds();
         auto status = InternalError.noerror;
         auto readExecResult = core.sys.posix.unistd.read(forkPipe[0], &status, status.sizeof);
+        // Save error number just in case if subsequent "waitpid" fails and overrides errno
+        auto lastError = .errno;
```
Comment (andralex) [inline std/process.d]: immutable

File: std/process.d [hunk 22]
```diff
@@ -571,15 +637,28 @@ private Pid spawnProcessImpl(in char[][] args,
...
+                    assert(config & Config.detached);
+                    errorMsg = "Failed to fork twice";
+                    break;
                 case InternalError.noerror:
                     assert(false);
             }
-
             if (readExecResult == error.sizeof)
                 throw ProcessException.newFromErrno(error, errorMsg);
             else
```
Comment (andralex) [inline std/process.d]: delete

File: std/process.d [hunk 23]
```diff
@@ -571,15 +637,28 @@ private Pid spawnProcessImpl(in char[][] args,
...
                 throw new ProcessException(errorMsg);
         }
+        else if (config & Config.detached)
+        {
+            owned = false;
+            typeof(id) actualPid;
+            if (read(pidPipe[0], &actualPid, id.sizeof) == id.sizeof)
+                id = actualPid;
+            else
+                throw ProcessException.newFromErrno("Could not read from pipe to get detached process id");
```
Comment (andralex) [inline std/process.d]: simplify flow

File: std/process.d [hunk 24]
```diff
@@ -689,8 +768,15 @@ private Pid spawnProcessImpl(in char[] commandLine,
...
     // close the thread handle in the process info structure
     CloseHandle(pi.hThread);
-
-    return new Pid(pi.dwProcessId, pi.hProcess);
+    if (config & Config.detached)
+    {
+        CloseHandle(pi.hProcess);
+        return new Pid(pi.dwProcessId);
+    }
+    else
```
Comment (andralex) [inline std/process.d]: felete

**Comment by andralex** [comment]
>move away from this single point of failure and redefine this policy to "at least two codeowners need to approve a new symbol" (or require the @andralex label only for modules he cares about, i.e. for which he is a code owner)

Please let me be the point of failure for the sake of better checks and balances of library consistency. In several prior cases it has happened that design sensibilities somewhat foreign to D/Phobos have entered the standard library, to everybody's displeasure. Thanks.

File: std/process.d (hunk 24 above)
Comment (schveiguy) [inline std/process.d]: delete? ;)

File: std/process.d [hunk 25]
```diff
@@ -1009,40 +1095,58 @@ version (Posix) @system unittest
...
+        std.file.write(pathi, "INPUT"~std.ascii.newline);
+        auto filei = File(pathi, "r");
+        auto fileo = File(patho, "w");
+        auto filee = File(pathe, "w");
+        auto pid = spawnProcess([prog.path, "bar", "baz" ], filei, fileo, filee, null, config);
+        if (!(config & Config.detached))
+            wait(pid);
+        else
+            // We need to wait a little to ensure that the process has finished and data was written to files
+            Thread.sleep(dur!"msecs"(200));
```
Comment (schveiguy) [inline std/process.d]: Wait longer. There's no need to vie for performance here. Wait 2 seconds instead of .2

File: std/process.d [hunk 26]
```diff
@@ -1712,6 +1849,25 @@ void kill(Pid pid, int codeOrSignal)
...
+    import std.exception : assertThrown;
+    TestScript prog = "exit 0";
+    auto pid = spawnProcess([prog.path], null, Config.detached);
+    /*
+    This sleep is needed because we can't wait() for detached process to end
+    and therefore TestScript destructor may run at the same time as /bin/sh tries to start the script.
+    This leads to the annoying message like "/bin/sh: 0: Can't open /tmp/std.process temporary file" to appear when running tests.
+    It does not happen in unittests with non-detached processes because we always wait() for them to finish.
+    */
+    Thread.sleep(dur!"msecs"(100));
```
Comment (schveiguy) [inline std/process.d]: Longer time here too.

File: std/process.d (hunk 25 above)
Comment (schveiguy) [inline std/process.d]: Also, stylistic change, you can put `200.msecs` instead of `dur!"msecs"(200)`

File: std/process.d (hunk 19 above)
Comment (FreeSlave) [inline std/process.d]: It was like this in previous implementation

File: std/process.d (hunk 20 above)
Comment (FreeSlave) [inline std/process.d]: Again, you're asking to change things that were like this before me. Well, I can, for sure.

File: std/process.d (hunk 24 above)
Comment (FreeSlave) [inline std/process.d]: Delete what? Can you elaborate?

File: std/process.d (hunk 10 above)
Comment (FreeSlave) [inline std/process.d]: Already considered in the comment before `int[2] pidPipe;`

File: std/process.d (hunk 24 above)
Comment (andralex) [inline std/process.d]: I mean delete the `else` line.

File: std/process.d (hunk 20 above)
Comment (andralex) [inline std/process.d]: Yes and much appreciated!

File: std/process.d (hunk 25 above)
Comment (FreeSlave) [inline std/process.d]: I would prefer unittests run as fast as they could. But ok.

**Comment by schveiguy** [comment]
ping @CyberShadow any idea why the DAutoTest fails?

**Comment by schveiguy** [comment]
> Bugzilla references

@wilzbach this looks invalid to me, this has nothing to do with RedBlackTree. I don't actually see a bugzilla reference in any commit, so not sure why this showed up.

**Comment by CyberShadow** [comment]
> ping @CyberShadow any idea why the DAutoTest fails?

Please see https://github.com/dlang/phobos/pull/5639#issuecomment-316838940 / https://github.com/dlang/phobos/pull/5626#issuecomment-316285183 / https://github.com/dlang/phobos/pull/5625#issuecomment-315921337 / https://github.com/dlang/phobos/pull/5619#issuecomment-315758105 / https://github.com/dlang/phobos/pull/5617#issuecomment-315668792.

TL;DR: GitHub bug.

**Comment by schveiguy** [comment]
@CyberShadow thanks, here's a followup: while this bug exists, what is the correct response, seeing as that test is required for merging? Should I just ping you/@wilzbach, or is there a way I can fix it myself?

**Comment by wilzbach** [comment]
> while this bug exists, what is the correct response, seeing as that test is required for merging? Should I just ping you/@wilzbach, or is there a way I can fix it myself?

AFAICT the only safe way is to create a new PR.
It also needs to be a different commit sha :/

We can overwrite it manually if the auto-tester passes, but there is always the slight risk of breaking Jenkins or CircleCI though. However, at least for CircleCi we get a warning mail afterwards, so I think we can start to risk it. Creating a new PR all the time, is really annoying.

**Comment by wilzbach** [comment]
> @wilzbach this looks invalid to me, this has nothing to do with RedBlackTree. I don't actually see a bugzilla reference in any commit, so not sure why this showed up.

I would guess that it's due to the GitHub API sending rubbish.

**Comment by wilzbach** [comment]
> We can overwrite it manually if the auto-tester passes, but there is always the slight risk of breaking Jenkins or CircleCI though. However, at least for CircleCi we get a warning mail afterwards, so I think we can start to risk it. Creating a new PR all the time, is really annoying.

Tried it this time - let's hope for the best. We really need to get GitHub to fix their bugs though..
